com.hazelcast.jet.core.metrics.MetricsTest.customUnit_notUsed [HZ-997]

_5.0.z_ (commit ce4ba622a7846f34def924fb13c5c3e613680523)

Failed on Windows, Oracle JDK8: [test report](http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-5.maintenance-Windows-OracleJDK8/18/testReport/junit/com.hazelcast.jet.core.metrics/MetricsTest/customUnit_notUsed/)

<details><summary>Stacktrace:</summary>

```
java.lang.AssertionError: Expected measurements for metric 'sum', but there were none
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at org.junit.Assert.assertFalse(Assert.java:65)
	at com.hazelcast.jet.core.metrics.JobMetricsChecker.assertAggregatedMetricValue(JobMetricsChecker.java:76)
	at com.hazelcast.jet.core.metrics.JobMetricsChecker.assertSummedMetricValue(JobMetricsChecker.java:54)
	at com.hazelcast.jet.core.metrics.MetricsTest.customUnit_notUsed(MetricsTest.java:164)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

</details>

<details><summary>Standard output:</summary>

```
0:01:00,933  INFO |customUnit| - [JetTestSupport] Thread-44565 - Terminating instanceFactory in JetTestSupport.@After
Finished Running Test: nonCooperativeProcessor in 0.713 seconds.
20:01:00,947 DEBUG |metricInFusedStages| - [JobCoordinationService] hz.determined_austin.cached.thread-4 - [127.0.0.1]:5701 [dev] [5.0.1-SNAPSHOT] Not starting jobs because partitions are not yet initialized.
20:01:00,947  INFO |customUnit| - [LifecycleService] Thread-44573 - [127.0.0.1]:5701 [dev] [5.0.1-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
20:01:00,948  WARN |customUnit| - [Node] Thread-44573 - [127.0.0.1]:5701 [dev] [5.0.1-SNAPSHOT] Terminating forcefully...
20:01:00,948  INFO |customUnit| - [Node] Thread-44573 - [127.0.0.1]:5701 [dev] [5.0.1-SNAPSHOT] Shutting down connection manager...
20:01:00,948  INFO |customUnit| - [Node] Thread-44573 - [127.0.0.1]:5701 [dev] [5.0.1-SNAPSHOT] Shutting down node engine...
```

</details>

--- a/hazelcast/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java
+++ b/hazelcast/src/main/java/com/hazelcast/jet/impl/JobCoordinationService.java
@@ -943,7 +943,7 @@ public class JobCoordinationService {
                             ? masterContext.jobContext().jobMetrics()
                             : null;
             jobRepository.completeJob(masterContext, jobMetrics, error, completionTime);
-            if (masterContexts.remove(masterContext.jobId(), masterContext)) {
+            if (removeMasterContext(masterContext)) {
                 completeObservables(masterContext.jobRecord().getOwnedObservables(), error);
                 logger.fine(masterContext.jobIdString() + " is completed");
                 (error == null ? jobCompletedSuccessfully : jobCompletedWithFailure).inc();
@@ -958,6 +958,12 @@ public class JobCoordinationService {
             }
             unscheduleJobTimeout(masterContext.jobId());
         });
+    }
+
+    private boolean removeMasterContext(MasterContext masterContext) {
+        synchronized (lock) {
+            return masterContexts.remove(masterContext.jobId(), masterContext);
+        }
     }
 
     /**
@@ -1102,15 +1108,20 @@ public class JobCoordinationService {
     ) {
         // the order of operations is important.
         long jobId = jobRecord.getJobId();
-        JobResult jobResult = jobRepository.getJobResult(jobId);
-        if (jobResult != null) {
-            logger.fine("Not starting job " + idToString(jobId) + ", already has result: " + jobResult);
-            return jobResult.asCompletableFuture();
-        }
 
         MasterContext masterContext;
         MasterContext oldMasterContext;
         synchronized (lock) {
+            // We check the JobResult while holding the lock to avoid this scenario:
+            // 1. We find no job result
+            // 2. Another thread creates the result and removes the master context in completeJob
+            // 3. We re-create the master context below
+            JobResult jobResult = jobRepository.getJobResult(jobId);
+            if (jobResult != null) {
+                logger.fine("Not starting job " + idToString(jobId) + ", already has result: " + jobResult);
+                return jobResult.asCompletableFuture();
+            }
+
             checkOperationalState();
 
             masterContext = createMasterContext(jobRecord, jobExecutionRecord);
@@ -1121,11 +1132,10 @@ public class JobCoordinationService {
             return oldMasterContext.jobContext().jobCompletionFuture();
         }
 
-        // If job is not currently running, it might be that it just completed.
-        // Since we've put the MasterContext into the masterContexts map, someone else could
-        // have joined to the job in the meantime so we should notify its future.
-        if (completeMasterContextIfJobAlreadyCompleted(masterContext)) {
-            return masterContext.jobContext().jobCompletionFuture();
+        assert jobRepository.getJobResult(jobId) == null : "jobResult should not exist at this point";
+
+        if (finalizeJobIfAutoScalingOff(masterContext)) {
+            masterContext.jobContext().jobCompletionFuture();
         }
 
         if (jobExecutionRecord.isSuspended()) {
@@ -1146,16 +1156,19 @@ public class JobCoordinationService {
             logger.fine("Completing master context for " + masterContext.jobIdString()
                     + " since already completed with result: " + jobResult);
             masterContext.jobContext().setFinalResult(jobResult.getFailureAsThrowable());
-            return masterContexts.remove(jobId, masterContext);
+            return removeMasterContext(masterContext);
         }
 
+        return finalizeJobIfAutoScalingOff(masterContext);
+    }
+
+    private boolean finalizeJobIfAutoScalingOff(MasterContext masterContext) {
         if (!masterContext.jobConfig().isAutoScaling() && masterContext.jobExecutionRecord().executed()) {
             logger.info("Suspending or failing " + masterContext.jobIdString()
                     + " since auto-restart is disabled and the job has been executed before");
             masterContext.jobContext().finalizeJob(new TopologyChangedException());
             return true;
         }
-
         return false;
     }
 

